src/providers/pdal/qgspdalprovider.cpp

```diff
@@ -0,0 +1,151 @@
+                         qgspdaldataprovider.cpp
+                         -----------------------
+  Date                 : November 2020
+  Copyright            : (C) 2020 by Peter Petrik
+  Email                : zilolv at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgis.h"
+#include "qgspdalprovider.h"
+#include "qgspdaldataitems.h"
+#include "qgsruntimeprofiler.h"
+#include "qgsapplication.h"
+
+#include <pdal/io/LasReader.hpp>
+#include <pdal/io/LasHeader.hpp>
+#include <pdal/Options.hpp>
+
+#define PROVIDER_KEY QStringLiteral( "pdal" )
+#define PROVIDER_DESCRIPTION QStringLiteral( "PDAL point cloud data provider" )
+
+QgsPdalProvider::QgsPdalProvider(
+  const QString &uri,
+  const QgsDataProvider::ProviderOptions &options,
+  QgsDataProvider::ReadFlags flags )
+  : QgsPointCloudDataProvider( uri, options, flags )
+{
+  std::unique_ptr< QgsScopedRuntimeProfile > profile;
+  if ( QgsApplication::profiler()->groupIsActive( QStringLiteral( "projectload" ) ) )
+    profile = qgis::make_unique< QgsScopedRuntimeProfile >( tr( "Open data source" ), QStringLiteral( "projectload" ) );
+
+  mIsValid = load( uri );
+}
+
+QgsPdalProvider::~QgsPdalProvider() = default;
+
+QgsCoordinateReferenceSystem QgsPdalProvider::crs() const
+{
+  return mCrs;
+}
+
+QgsRectangle QgsPdalProvider::extent() const
+{
+  return mExtent;
+}
+
+bool QgsPdalProvider::isValid() const
+{
+  return mIsValid;
+}
+
+QString QgsPdalProvider::name() const
+{
+  return QStringLiteral( "pdal" );
+}
+
+QString QgsPdalProvider::description() const
+{
+  return QStringLiteral( "Point Clouds PDAL" );
+}
+
+QgsPointCloudIndex *QgsPdalProvider::index() const
+{
+  // TODO automatically generate EPT index
+  return nullptr;
+}
+
+bool QgsPdalProvider::load( const QString &uri )
+{
+  pdal::Option las_opt( "filename", uri.toStdString() );
+  pdal::Options las_opts;
+  las_opts.add( las_opt );
+  pdal::PointTable table;
+  pdal::LasReader las_reader;
+  las_reader.setOptions( las_opts );
+  las_reader.prepare( table );
+  pdal::PointViewSet point_view_set = las_reader.execute( table );
+  pdal::PointViewPtr point_view = *point_view_set.begin();
+  pdal::Dimension::IdList dims = point_view->dims();
+  pdal::LasHeader las_header = las_reader.header();
```

Comment:
Is this part relatively stable? (ie. not expected to change before release?) If so, I can add the code required to read and auto populate the layer metadata from this LasHeader object.

Reply:
not sure how @wonder-sk want to approach it. Do we support "ALL" PDAL supported formats? in that case do we need some QGIS abstraction to header/readers (this one is LasHeader -> only for laz/las files). 
Detto for the input source dialog. For EPT format it is easy. For PDAL-supported formats you may need to select (manually) driver in the dialog. You may want to select if you you want/do not want to automatically index it. Also you may want to add JSON file with the PDAL filters to let PDAL/EPT know how to read/index it... etc.

Reply:
I still think we can avoid two seperate source select widgets. I'm thinking of an approach like what is done for vector layers now, where format specific options are shown after picking the input file. But I'll do handle this after this pr is merged, I've got a good idea of exactly how to handle this.

Reply:
great!